File: android/src/main/kotlin/com/usesense/flutter/UseSenseFlowsBridge.kt
package com.usesense.flutter

import android.app.Activity
import android.os.Handler
import android.os.Looper
import com.usesense.sdk.flows.FlowError
import com.usesense.sdk.flows.FlowOutcome
import com.usesense.sdk.flows.FlowRunResult
import com.usesense.sdk.flows.FlowRunState
import com.usesense.sdk.flows.FlowsCallback
import com.usesense.sdk.flows.UseSenseFlows
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * MethodChannel bridge for the Flows runner. Sits alongside the existing
 * Pigeon-generated Sessions surface, sidestepping pigeon regen. Channel
 * name: `com.usesense.flutter/flows`. Single inbound method: `runFlow`.
 *
 * On success the result is a Map<String, Any?> with flowRunId/state/outcome;
 * on failure the result is a FlutterError with the FlowError code as `code`.
 */
class UseSenseFlowsBridge {
    private val mainHandler = Handler(Looper.getMainLooper())

    fun handle(call: MethodCall, result: MethodChannel.Result, activity: Activity?) {
        when (call.method) {
            "runFlow" -> {
                val flowRunId = call.argument<String>("flowRunId")
                val sdkToken = call.argument<String>("sdkToken")
                val apiBaseUrl = call.argument<String>("apiBaseUrl") ?: "https://api.usesense.ai"
                if (flowRunId == null || sdkToken == null) {
                    result.error("unknown", "flowRunId and sdkToken are required", null)
                    return
                }
                val act = activity ?: run {
                    result.error("unknown", "No current activity to launch from", null)
                    return
                }

                val callback = object : FlowsCallback {
                    override fun onResult(r: FlowRunResult) {
                        mainHandler.post {
                            result.success(
                                mapOf(
                                    "flowRunId" to r.flowRunId,
                                    "state" to stateWire(r.state),
                                    "outcome" to r.outcome?.let { outcomeWire(it) },
                                ),
                            )
                        }
                    }

                    override fun onError(error: FlowError) {
                        mainHandler.post {
                            result.error(error.code.wire, error.message, null)
                        }
                    }
                }

                UseSenseFlows.run(
                    activity = act,
                    flowRunId = flowRunId,
                    sdkToken = sdkToken,
                    callback = callback,
                    apiBaseUrl = apiBaseUrl,
                )
            }
            else -> result.notImplemented()
        }
    }

    private fun stateWire(state: FlowRunState): String = state.wire

    private fun outcomeWire(outcome: FlowOutcome): String = outcome.wire
}

File: ios/Classes/UseSenseFlowsBridge.swift
import Flutter
import UIKit
import UseSenseSDK

/// MethodChannel bridge for the Flows runner. Sits alongside the existing
/// Pigeon-generated Sessions surface, sidestepping pigeon regen. Channel
/// name: `com.usesense.flutter/flows`. Single inbound method: `runFlow`.
///
/// On success: the Future resolves with a Map containing flowRunId/state/outcome.
/// On failure: a FlutterError with the FlowError code as `code`.
final class UseSenseFlowsBridge: NSObject {
    private weak var registrar: FlutterPluginRegistrar?

    init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
        super.init()
    }

    /// Handle a single MethodChannel invocation. Dispatched on the main
    /// thread by Flutter so the SDK's UIKit presentation works directly.
    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "runFlow":
            guard let args = call.arguments as? [String: Any],
                  let flowRunId = args["flowRunId"] as? String,
                  let sdkToken = args["sdkToken"] as? String else {
                result(FlutterError(code: "unknown", message: "flowRunId and sdkToken are required", details: nil))
                return
            }
            let apiBaseURLString = (args["apiBaseUrl"] as? String) ?? "https://api.usesense.ai"
            guard let apiBaseURL = URL(string: apiBaseURLString) else {
                result(FlutterError(code: "unknown", message: "Invalid apiBaseUrl", details: nil))
                return
            }
            guard let presenter = Self.topViewController() else {
                result(FlutterError(code: "unknown", message: "No view controller to present from", details: nil))
                return
            }

            UseSenseFlows.run(
                flowRunId: flowRunId,
                sdkToken: sdkToken,
                apiBaseURL: apiBaseURL,
                from: presenter,
            ) { runResult in
                switch runResult {
                case .success(let r):
                    result([
                        "flowRunId": r.flowRunId,
                        "state": Self.stateWire(r.state),
                        "outcome": r.outcome.map(Self.outcomeWire) as Any,
                    ])
                case .failure(let e):
                    result(FlutterError(code: e.code.rawValue, message: e.message, details: nil))
                }
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Helpers

    private static func stateWire(_ state: FlowRunState) -> String {
        switch state {
        case .pending: return "pending"
        case .inProgress: return "in_progress"
        case .stalled: return "stalled"
        case .awaitingReview: return "awaiting_review"
        case .completed: return "completed"
        case .errored: return "errored"
        case .abandoned: return "abandoned"
        case .cancelled: return "cancelled"
        }
    }

    private static func outcomeWire(_ outcome: FlowOutcome) -> String {
        switch outcome {
        case .approve: return "APPROVE"
        case .reject: return "REJECT"
        case .manualReview: return "MANUAL_REVIEW"
        }
    }

    /// Find the top-most view controller to present from. Mirrors the
    /// pattern the V4 bridge uses; works across UIKit + SceneDelegate apps
    /// and falls back gracefully to `keyWindow.rootViewController`.
    private static func topViewController() -> UIViewController? {
        var root: UIViewController?
        if #available(iOS 13.0, *) {
            for scene in UIApplication.shared.connectedScenes {
                guard let windowScene = scene as? UIWindowScene else { continue }
                if let keyWindow = windowScene.windows.first(where: { $0.isKeyWindow }) ?? windowScene.windows.first {
                    root = keyWindow.rootViewController
                    break
                }
            }
        }
        if root == nil {
            root = UIApplication.shared.keyWindow?.rootViewController
        }
        var current = root
        while let presented = current?.presentedViewController {
            current = presented
        }
        return current
    }
}
